server/app.py

```diff
@@ -266,50 +262,80 @@ def player_finished(data):
 
     return ROOMS[room]['playersFinished']
 
 @SOCKETIO.on('goBackToLobby')
 def go_back_to_lobby(data):
-    '''
-    This function is called when a player goes back to their lobby.
-    All of the players will follow and 'playersFinished' will be cleared.
-    '''
+    '''This function is called when a player goes back to their lobby.
+    All of the players will follow and 'playersFinished' will be cleared'''
     room = data['room']
     ROOMS[room]['playersFinished'].clear()
     SOCKETIO.emit('goBackToLobby', include_self=False, broadcast=True, room=room)
 
+
+@SOCKETIO.on('leaderboard')
+def on_leaderboard_query(data):
+    '''This function is called when the leaderboard is being rendered. It fetches data from
+    the database and then calculates everyones avg wpm and send all the info back to the client'''
+    if data["sort_query"] != "avgwpm":
+        db_usersnames, db_emails = fetch_db(data["sort_query"])[0:2]
+        db_bestwpm, db_totalwpm, db_gamesplayed, db_gameswon = fetch_db(data["sort_query"])[3:]
+        calculate_avg = find_average(db_totalwpm, db_gamesplayed)
+        db_avgwpm = calculate_avg
+    else:
+        db_usersnames, db_emails = fetch_db(data["sort_query"])[0:2]
+        db_bestwpm, db_totalwpm, db_gamesplayed, db_gameswon = fetch_db(data["sort_query"])[3:]
+        calculated_avg = find_average(db_totalwpm, db_gamesplayed)
+        data = sort_avg_wpm(db_usersnames, db_emails, db_bestwpm,
+                            calculated_avg, db_gamesplayed, db_gameswon)
+        db_usersnames, db_emails, db_bestwpm, db_avgwpm, db_gamesplayed, db_gameswon = data
+    SOCKETIO.emit('updateLeaderboard', {'db_usersnames': db_usersnames,
+                                        'db_bestwpm': db_bestwpm,
+                                        'db_avgwpm': db_avgwpm,
+                                        'db_gamesplayed': db_gamesplayed,
+                                        'db_gameswon': db_gameswon},
+                  broadcast=True,
+                  include_self=True)
+
+
 def fetch_db(sort_by):
-    """This is how we fetch all of the information from Heroku's DB, it also allows us to order
-    the information by best wpm/emails(alphabetical)/gameswon"""
-    if sort_by == "wpm":
+    '''This is how we fetch all of the information from Heroku's DB, it also allows us to order
+    the information by best wpm/emails(alphabetical)/gameswon'''
+    if sort_by == "bestwpm":
         all_users = DB.session.query(models.Users).order_by(
             models.Users.bestwpm.desc()).all()
         return fetch_db_helper(all_users)
 
+    if sort_by == "gamesplayed":
+        all_users = DB.session.query(models.Users).order_by(
+            models.Users.gamesplayed.desc()).all()
+        return fetch_db_helper(all_users)
+
+    if sort_by == "gameswon":
+        all_users = DB.session.query(models.Users).order_by(
+            models.Users.gameswon.desc()).all()
```

Comment:
Do we need to `return fetch_db_helper(all_users)` here as well?

Reply:
We dont need it as im returning it at the end of the function anyways, But I will add it to keep it unifrom with all the other if statements!